spring-cloud-gcp-pubsub/src/test/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubBatchMessageSourceTests.java

```diff
@@ -0,0 +1,250 @@
+	}
+
+	@Test
+	@SuppressWarnings("unchecked")
+	public void doReceive_callsPubsubAndCachesCorrectly() {
+		when(this.mockPubSubSubscriberOperations.pull("sub1", 3, true))
+				.thenReturn(Arrays.asList(this.msg1, this.msg2, this.msg3));
+		PubSubBatchMessageSource pubSubMessageSource = new PubSubBatchMessageSource(
+				this.mockPubSubSubscriberOperations, "sub1");
+		pubSubMessageSource.setMaxFetchSize(3);
+		Message<Object> message = pubSubMessageSource.receive();
+		List<AcknowledgeablePubsubMessage> payload = (List<AcknowledgeablePubsubMessage>) message.getPayload();
+		assertThat(payload.size()).isEqualTo(3);
+		assertThat(payload.containsAll(Arrays.asList(msg1, msg2, msg3))).isTrue();
+	}
+
+	@Test
+	@SuppressWarnings("unchecked")
+	public void doReceive_pullsOneAtATimeWhenMaxFetchSizeNegative() {
+		PubSubBatchMessageSource pubSubMessageSource = new PubSubBatchMessageSource(
+				this.mockPubSubSubscriberOperations, "sub1");
+
+		MessageBuilder<AcknowledgeablePubsubMessage> message = (MessageBuilder<AcknowledgeablePubsubMessage>) pubSubMessageSource
+				.doReceive(-1);
+
+		assertThat(message).isNotNull();
+
+		verify(this.mockPubSubSubscriberOperations, times(1))
+				.pull("sub1", 1, true);
+	}
+
+	@Test
+	@SuppressWarnings("unchecked")
+	public void doReceive_manualAckModeAppliesAcknowledgmentHeaderAndDoesNotAck() {
+
+		PubSubBatchMessageSource pubSubMessageSource = new PubSubBatchMessageSource(
+				this.mockPubSubSubscriberOperations, "sub1");
+		pubSubMessageSource.setMaxFetchSize(1);
+		pubSubMessageSource.setAckMode(AckMode.MANUAL);
+
+		Message<Object> message = pubSubMessageSource.receive();
+		assertThat(message).isNotNull();
+		List<AcknowledgeablePubsubMessage> payload = (List<AcknowledgeablePubsubMessage>) message.getPayload();
+
+		assertThat(payload).containsOnly(msg1);
+
+		PubSubBatchAcknowledgmentCallback callback = (PubSubBatchAcknowledgmentCallback) message.getHeaders()
+				.get(IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK);
+		assertThat(callback).isNotNull();
+		assertThat(callback.isAcknowledged()).isFalse();
+		verify(this.msg1, times(0)).ack();
+
+		callback.acknowledge(AcknowledgmentCallback.Status.ACCEPT);
+		verify(this.msg1, times(1)).ack();
+		assertThat(callback.isAcknowledged()).isTrue();
+	}
+
+	@Test
+	public void doReceive_autoModeAcksAndAddsOriginalMessageHeader() {
+
+		PubSubBatchMessageSource pubSubMessageSource = new PubSubBatchMessageSource(
+				this.mockPubSubSubscriberOperations, "sub1");
+		pubSubMessageSource.setMaxFetchSize(1);
+		pubSubMessageSource.setAckMode(AckMode.AUTO);
+		MessageSourcePollingTemplate poller = new MessageSourcePollingTemplate(pubSubMessageSource);
+		poller.poll((message) -> {
+			assertThat(message).isNotNull();
+			List<AcknowledgeablePubsubMessage> payload = (List<AcknowledgeablePubsubMessage>) message.getPayload();
+			assertThat(payload).containsOnly(msg1);
+			assertThat(message.getHeaders()).doesNotContainKey(GcpPubSubHeaders.ORIGINAL_MESSAGE);
```

Comment:
The name of the method contradicts this assertion. You probably want to check for the `ACKNOWLEDGMENT_CALLBACK` instead?